www/__tests__/enketoHelper.test.ts

```diff
@@ -0,0 +1,245 @@
+  const filled =
+    '<aDxjD5f5KAghquhAvsormy xmlns:jr="http://openrosa.org/javarosa" xmlns:orx="http://openrosa.org/xforms" id="time_use_survey_form_v9_1"><start/><end/><group_hg4zz25><Start_date>2016-07-25</Start_date><Start_time>17:24:32.928-06:00</Start_time><End_date>2016-07-25</End_date><End_time>17:30:31.000-06:00</End_time><Activity_Type/><Personal_Care_activities/><Employment_related_a_Education_activities/><Domestic_activities/><Recreation_and_leisure/><Voluntary_work_and_care_activities/><Other/></group_hg4zz25><meta><instanceID/></meta></aDxjD5f5KAghquhAvsormy>';
+  const opts = { prefilledSurveyResponse: filled };
+  const opts2 = {
+    prefillFields: {
+      Start_date: '2016-07-25',
+      Start_time: '17:24:32.928-06:00',
+      End_date: '2016-07-25',
+      End_time: '17:30:31.000-06:00',
+    },
+  };
+
+  //if no xmlModel, returns null
+  expect(getInstanceStr(null, opts)).toBe(null);
+  //if there is a prefilled survey, return it
+  expect(getInstanceStr(xmlModel, opts)).toBe(filled);
+  //if there is a model and fields, return prefilled
+  expect(getInstanceStr(xmlModel, opts2)).toBe(filled);
+  //if none of those things, also return null
+  expect(getInstanceStr(xmlModel, {})).toBe(null);
+});
+
+//resolve timestamps
+it('resolves the timestamps', () => {
+  const xmlParser = new window.DOMParser();
+  const timelineEntry = {
+    end_local_dt: { timezone: 'America/Los_Angeles' },
+    start_ts: 1469492672.928242,
+    end_ts: 1469493031,
+  };
+
+  //missing data returns null
+  const missingData =
+    '<tag> <Start_date>2016-08-28</Start_date> <End_date>2016-07-25</End_date> <End_time>17:30:31.000-06:00</End_time> </tag>';
+  const missDataDoc = xmlParser.parseFromString(missingData, 'text/html');
+  expect(resolveTimestamps(missDataDoc, timelineEntry)).toBeNull();
+  //bad time returns undefined
+  const badTimes =
+    '<tag> <Start_date>2016-08-28</Start_date> <End_date>2016-07-25</End_date> <Start_time>17:32:32.928-06:00</Start_time> <End_time>17:30:31.000-06:00</End_time> </tag>';
+  const badTimeDoc = xmlParser.parseFromString(badTimes, 'text/xml');
+  expect(resolveTimestamps(badTimeDoc, timelineEntry)).toBeUndefined();
+  //good info returns unix start and end timestamps -- TODO : address precise vs less precise?
+  const timeSurvey =
+    '<tag> <Start_date>2016-07-25</Start_date> <End_date>2016-07-25</End_date> <Start_time>17:24:32.928-06:00</Start_time> <End_time>17:30:31.000-06:00</End_time> </tag>';
+  const xmlDoc = xmlParser.parseFromString(timeSurvey, 'text/xml');
+  expect(resolveTimestamps(xmlDoc, timelineEntry)).toMatchObject({
+    start_ts: 1469492672.928242,
+    end_ts: 1469493031,
+  });
```

Comment:
I want to make sure one of the test cases is covered here:

When times from the survey response are within 1 minute of the timelineEntry's start_ts/end_ts, it should use the timelineEntry's start_ts/end_ts, otherwise it should use the times from the survey response.

Is that what is being tested here? It's a little hard to tell because start_ts, end_ts are in seconds and Enketo times are in HH:MM:SS

Reply:
I'm not sure I knew there was two different cases for what timestamps to use, I did write these tests a while back though, I will verify that I test both cases (timelineEntry and survey response) and add comments to clarify the cases

Reply:
I had noted that, but had not tested it yet, this has now been tested (and commented as such) [8f7c674](https://github.com/e-mission/e-mission-phone/pull/1063/commits/8f7c6748ed161e248e247cc04514dacc46908317)